middleware_rate_limiting.go

```diff
@@ -76,32 +76,52 @@ func (k *RateLimitAndQuotaCheck) handleQuotaFailure(r *http.Request, authHeaderV
 	return errors.New("Quota exceeded"), 403
 }
 
+func (k *RateLimitAndQuotaCheck) doSessionWrite(r *http.Request, authHeaderValue string, thisSessionState *SessionState) {
+	if !config.UseAsyncSessionWrite {
+		k.Spec.SessionManager.UpdateSession(authHeaderValue, *thisSessionState, GetLifetime(k.Spec, thisSessionState))
+		context.Set(r, SessionData, *thisSessionState)
+	} else {
+		go k.Spec.SessionManager.UpdateSession(authHeaderValue, *thisSessionState, GetLifetime(k.Spec, thisSessionState))
+		go context.Set(r, SessionData, *thisSessionState)
+	}
+}
+
 // ProcessRequest will run any checks on the request on the way through the system, return an error to have the chain fail
 func (k *RateLimitAndQuotaCheck) ProcessRequest(w http.ResponseWriter, r *http.Request, configuration interface{}) (error, int) {
 	sessionState := context.Get(r, SessionData).(SessionState)
 	authHeaderValue := context.Get(r, AuthHeaderValue).(string)
 
 	storeRef := k.Spec.SessionManager.GetStore()
+
+	oldQuotaRenewal := sessionState.QuotaRenews
 	forwardMessage, reason := sessionLimiter.ForwardMessage(&sessionState,
 		authHeaderValue,
 		storeRef,
 		!k.Spec.DisableRateLimit,
 		!k.Spec.DisableQuota)
 
-	// If either are disabled, save the write roundtrip
-	if !k.Spec.DisableRateLimit || !k.Spec.DisableQuota {
-		// Ensure quota and rate data for this session are recorded
-		if !config.UseAsyncSessionWrite {
-			k.Spec.SessionManager.UpdateSession(authHeaderValue, sessionState, GetLifetime(k.Spec, &sessionState))
-			context.Set(r, SessionData, sessionState)
-		} else {
-			go k.Spec.SessionManager.UpdateSession(authHeaderValue, sessionState, GetLifetime(k.Spec, &sessionState))
-			go context.Set(r, SessionData, sessionState)
+	// If either are disabled, or the session has been modified, save the write roundtrip
+	if k.Spec.DisableRateLimit == false || k.Spec.DisableQuota == false {
+		// If we are using the distributed counter, we don;t need to write data here,
+		// the flusher will take care of it
+		if config.UseDistributedQuotaCounter == false {
+			// Ensure quota and rate data for this session are recorded
+			k.doSessionWrite(r, authHeaderValue, &sessionState)
+		} else if sessionState.QuotaRenews != oldQuotaRenewal {
+			k.doSessionWrite(r, authHeaderValue, &sessionState)
```

Comment:
Going to leave this, it's tempting to merge them, but they deal with different configuration cases.